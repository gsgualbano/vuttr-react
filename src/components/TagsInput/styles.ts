import styled from 'styled-components';
import { ReactComponent as Icon } from '~/assets/icons/Icon-Close-2px.svg';
import { Theme } from '~/styles/global';

export const Container = styled.div`
  display: flex;
  flex-direction: row;
  flex-flow: row wrap;
  min-width: 100%;
  border: 1px solid;
  border-color: ${({ theme }: Theme) => (theme.colors.InputBorderColor ? theme.colors.InputBorderColor : '#ebeaed')};
  background: ${({ theme }: Theme) => (theme.colors.InputBackground ? theme.colors.InputBackground : '#f5f4f6')};
  border-radius: 5px;
`;

export const TextInput = styled.input.attrs({
  type: 'text',
})`
  background: inherit;
  border: none;
  padding: 10px;
  outline: none;
  width: 100%;
`;

export const RemoveButton = styled.button.attrs({
  type: 'button',
})`
  border: none;
  background: inherit;
  border-radius: inherit;
`;

export const IconContainer = styled.div`
  width: 100%;
  justify-content: center;
  align-items: center;
`;

export const CloseIcon = styled(Icon)`
  width: 8px;
  height: 8px;
  margin-right: 5px;

  g > path.a {
    stroke: #ffffff;
    stroke-width: 15px;
  }
`;

export const Text = styled.p`
  font-size: 16px;
  font-family: 'Source Sans Pro', sans-serif;
  font-weight: 600;
  text-align: center;
  margin: 2px;
`;
