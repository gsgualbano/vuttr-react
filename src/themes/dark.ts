export default {
  colors: {
    background: '#1B1A1A',
    textColor: '#fff',
    cardBackground: '#29292B',
    cardBorderColor: '#29292B',
    modalBackground: '#29292B',
    modalBorderColor: '#29292B',
    linkColor: '#ae8fd6',
    linkVisitedColor: '##8050bf',
    buttonOutlineBackground: '#29292B',
    InputBackground: '#9e9fa2',
    InputBorderColor: '#9e9fa2',
  },
};
