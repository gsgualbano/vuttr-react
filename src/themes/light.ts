export default {
  colors: {
    background: '#fcfcfd',
    textColor: '#170c3a',
    cardBackground: '#fcfcfd',
    cardBorderColor: '#ebeaed',
    modalBackground: '#fefefe',
    modalBorderColor: '#888',
    linkColor: 'rgb(0, 0, 238)',
    linkVisitedColor: 'rgb(85, 26, 139);',
    buttonOutlineBackground: '#ffffff',
    InputBackground: '#f5f4f6',
    InputBorderColor: '#ebeaed',
  },
};
