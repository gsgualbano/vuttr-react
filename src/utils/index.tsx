/* eslint-disable max-len */
import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

import { render } from '@testing-library/react';

import { ThemeProvider } from 'styled-components';
import { reducer } from '../store/ducks/tools';

interface Theme {
  colors: {
    background: string;
    textColor: string;
    cardBackground: string;
    cardBorderColor: string;
    modalBackground: string;
    modalBorderColor: string;
    linkColor: string;
    linkVisitedColor: string;
    buttonOutlineBackground: string;
    InputBackground: string;
    InputBorderColor: string;
  };
}

/** redux mock to render components that depends on the redux */
export const renderWithRedux = (
  component: JSX.Element,
  { initialState, store = createStore(reducer, initialState) }: any = {},
) => ({
  ...render(<Provider store={store}>{component}</Provider>),
  store,
});

/** render components thats depends on the theme provider */
export const renderWithTheme = (component: JSX.Element, theme: Theme) => render(<ThemeProvider theme={theme}>{component}</ThemeProvider>);

/** render components thats depends on the redux and theme provider */
export const renderWithReduxAndTheme = (
  component: JSX.Element,
  theme: Theme,
  { initialState, store = createStore(reducer, initialState) }: any = {},
) => ({
  ...render(
    <ThemeProvider theme={theme}>
      <Provider store={store}>{component}</Provider>
    </ThemeProvider>,
  ),
  store,
});
