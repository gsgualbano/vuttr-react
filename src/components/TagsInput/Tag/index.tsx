import React from 'react';

import { Container } from './styles';

/**
 * Propriedades da Tag
 */
interface OwnProps {
  /** elemento jsx */
  children: React.ReactNode;
}

/**
 * Tag badge.
 */
const Tag = ({ children }: OwnProps): JSX.Element => <Container>{children}</Container>;

export default Tag;
