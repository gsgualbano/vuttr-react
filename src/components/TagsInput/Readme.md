TagsInput exemplo:

```js
import { useState } from "react";

const [tags, setTags] = useState([]);

<TagsInput id='tags' name='tags' tags={tags} setTags={setTags} />;
```

TagsInput Dark exemplo:

```js
import { useState } from "react";
import { ThemeProvider } from "styled-components";
import dark from "~/themes/dark";

const [tags, setTags] = useState([]);

<ThemeProvider theme={dark}>
  <>
    <TagsInput id='tags' name='tags' tags={tags} setTags={setTags} />
  </>
</ThemeProvider>;
```
