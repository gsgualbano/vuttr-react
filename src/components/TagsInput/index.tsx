import React, { useState } from 'react';

import {
 Container, TextInput, RemoveButton, IconContainer, CloseIcon, Text,
} from './styles';
import Tag from './Tag';

/**
 * Propreidades do TagsInput
 */
interface OwnProps {
  /** text input Id */
  id: string;
  /** text input name */
  name: string;
  /** tags array */
  tags: string[];
  /** dispatcher para atualizar as tags */
  setTags: React.Dispatch<React.SetStateAction<string[]>>;
}

/**
 * TextInput para adicionar tags.
 */
const TagsInput = ({
 id, name, tags, setTags,
}: OwnProps): JSX.Element => {
  // state
  const [text, setText] = useState<string>('');

  // TextInput change
  const handleOnChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setText(e.target.value);
  };

  // On Press of a Key on the TextInput
  const handleKeyDown = (e: React.KeyboardEvent): void => {
    if (e.keyCode === 32) {
      e.preventDefault();
      setText('');
      setTags([...tags, text]);
    }
  };

  // On the click of the remove icon
  const handleOnClick = (i: number): void => {
    if (i !== -1) {
      setTags((tagsArr: string[]) => tagsArr.filter((_, index) => index !== i));
    }
  };

  return (
    <Container data-testid="tags-input">
      {tags.map((tag: string, i: number) => (
        <Tag key={tag}>
          <Text>{tag}</Text>
          <RemoveButton onClick={() => handleOnClick(i)} data-testid={`remove-${tag}`}>
            <IconContainer>
              <CloseIcon />
            </IconContainer>
          </RemoveButton>
        </Tag>
      ))}
      <TextInput
        id={id}
        name={name}
        value={text}
        placeholder="Add tag and press space"
        onKeyDown={handleKeyDown}
        onChange={handleOnChange}
      />
    </Container>
  );
};

export default TagsInput;
