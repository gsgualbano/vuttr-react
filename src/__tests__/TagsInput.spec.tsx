import React from 'react';
import { cleanup, fireEvent } from '@testing-library/react';
import { renderWithTheme } from '~/utils';

import TagsInput from '~/components/TagsInput';

import 'jest-dom/extend-expect';
import 'jest-styled-components';

import dark from '~/themes/dark';
import light from '~/themes/light';

afterEach(cleanup);

describe('<TagsInput />', () => {
  it('render correctly', () => {
    const { asFragment } = renderWithTheme(
      <TagsInput id="tags" name="tags" tags={['google', 'search', 'engine']} setTags={() => {}} />,
      light,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('render tags', () => {
    const { queryByText } = renderWithTheme(
      <TagsInput id="tags" name="tags" tags={['google', 'search', 'engine']} setTags={() => {}} />,
      light,
    );
    expect(queryByText('google')).toBeInTheDocument();
    expect(queryByText('search')).toBeInTheDocument();
    expect(queryByText('engine')).toBeInTheDocument();
  });

  it('have correct color light theme', () => {
    const { getByTestId } = renderWithTheme(
      <TagsInput id="tags" name="tags" tags={['google', 'search', 'engine']} setTags={() => {}} />,
      light,
    );
    expect(getByTestId('tags-input')).toHaveStyleRule('background', '#f5f4f6');
    expect(getByTestId('tags-input')).toHaveStyleRule('border-color', '#ebeaed');
  });

  it('have correct color dark theme', () => {
    const { getByTestId } = renderWithTheme(
      <TagsInput id="tags" name="tags" tags={['google', 'search', 'engine']} setTags={() => {}} />,
      dark,
    );
    expect(getByTestId('tags-input')).toHaveStyleRule('background', '#9e9fa2');
    expect(getByTestId('tags-input')).toHaveStyleRule('border-color', '#9e9fa2');
  });
});
