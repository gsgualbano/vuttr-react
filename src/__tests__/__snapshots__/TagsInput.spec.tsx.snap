// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TagsInput /> render correctly 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  min-width: 100%;
  border: 1px solid;
  border-color: #ebeaed;
  background: #f5f4f6;
  border-radius: 5px;
}

.c6 {
  background: inherit;
  border: none;
  padding: 10px;
  outline: none;
  width: 100%;
}

.c3 {
  border: none;
  background: inherit;
  border-radius: inherit;
}

.c4 {
  width: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 {
  width: 8px;
  height: 8px;
  margin-right: 5px;
}

.c5 g > path.a {
  stroke: #ffffff;
  stroke-width: 15px;
}

.c2 {
  font-size: 16px;
  font-family: 'Source Sans Pro',sans-serif;
  font-weight: 600;
  text-align: center;
  margin: 2px;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  background: #f95e5a;
  border-radius: 13px;
  padding: 2px;
  margin: 2px;
  color: #ffffff;
}

<div
    class="c0"
    data-testid="tags-input"
  >
    <div
      class="c1"
    >
      <p
        class="c2"
      >
        google
      </p>
      <button
        class="c3"
        data-testid="remove-google"
        type="button"
      >
        <div
          class="c4"
        >
          <svg
            class="c5"
          >
            Icon-Close-2px.svg
          </svg>
        </div>
      </button>
    </div>
    <div
      class="c1"
    >
      <p
        class="c2"
      >
        search
      </p>
      <button
        class="c3"
        data-testid="remove-search"
        type="button"
      >
        <div
          class="c4"
        >
          <svg
            class="c5"
          >
            Icon-Close-2px.svg
          </svg>
        </div>
      </button>
    </div>
    <div
      class="c1"
    >
      <p
        class="c2"
      >
        engine
      </p>
      <button
        class="c3"
        data-testid="remove-engine"
        type="button"
      >
        <div
          class="c4"
        >
          <svg
            class="c5"
          >
            Icon-Close-2px.svg
          </svg>
        </div>
      </button>
    </div>
    <input
      class="c6"
      id="tags"
      name="tags"
      placeholder="Add tag and press space"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
